Add sleep, passive mode for sensor

serunis opened the pull request:
Hello, I add sleep and passive mode for low power.
Tested with PMS7003 on windows10 Desktop, Raspberry pi 3B

Although I test only PMS7003, this would work for PMS5003 too.

Thank you

serunis committed:
add sleep, passive mode for sensor

Signed-off-by: serunis <53402902+serunis@users.noreply.github.com>

FlorentinBulotAQ commented:
Dear Serunis,

Thanks a lot for your pull request. I think it makes much sense to add these features. 
I have seen on the other closed pull request that you found one other item to fix so let us know if that is still the case and when you have fixed and and we will review your changes.

With best wishes,
Florentin.

serunis commented:
Hello Florentin,

I fixed the missing point, and updated with 'git push --force', so you cannot see the history,

I fixed the following flushing point:

    def read_in_passive(self, perform_flush=True):
        if perform_flush:
            self.serial.reset_input_buffer()  <-- 1. input flush here
        pms_cmd_read_in_passive = b'\x42\x4d\xe2\x00\x00\x01\x71'
        self.serial.write(pms_cmd_read_in_passive)
        self.serial.flush()
        ret = self.read(False)      <-- 2. call with False

My pull request is ready for review.
Thank you.

pjb304 commented:
Hi Serunis,
Many thanks for your contribution to this.
Flo is better placed to review this than I am so I've changed the reviewer over.

Phil

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+            immediate reading sensor data gives you 0 values.
+        """
+
+        pms_cmd_changemode_passive = b'\x42\x4d\xe1\x00\x00\x01\x70'
```
I would suggest you put these two variables (pms_cmd_changemode_passive and pms_cmd_changemode_active) as global variable for consistency.

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+    def read_in_passive(self, perform_flush=True):
+        if perform_flush:
+            self.serial.reset_input_buffer()  #Flush any data in the buffer
+        pms_cmd_read_in_passive = b'\x42\x4d\xe2\x00\x00\x01\x71'
```
Global variable

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+            From datasheet "Stable data should be got at least 30 seconds after the sensor wakeup
+            from the sleep mode because of the fan's performance."
+        """
+        pms_cmd_to_sleep = b'\x42\x4d\xe4\x00\x00\x01\x73'
```
Global variable

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+            from the sleep mode because of the fan's performance."
+        """
+        pms_cmd_to_sleep = b'\x42\x4d\xe4\x00\x00\x01\x73'
+        pms_cmd_to_wakeup = b'\x42\x4d\xe4\x00\x01\x01\x74'
```
global variable

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+        else:
+            self.serial.write(pms_cmd_changemode_active)
+
+        self.serial.flush()
```
I suggest using `self.serial.reset_input_buffer()` instead of `self.serial.flush()`

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+            self.serial.reset_input_buffer()  #Flush any data in the buffer
+        pms_cmd_read_in_passive = b'\x42\x4d\xe2\x00\x00\x01\x71'
+        self.serial.write(pms_cmd_read_in_passive)
+        self.serial.flush()
```
I suggest using `self.serial.reset_input_buffer()` instead of `self.serial.flush()`

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+        else:
+            self.serial.write(pms_cmd_to_wakeup)
+
+        self.serial.flush()
```
Use `self.serial.reset_input_buffer()` instead of `self.serial.flush()`

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+        return
+
+
+"""
```
This test script was very useful to test your code.

Could you put this test script into a separate file test.py ?

serunis replied:
I don't know how to make test codes in the pythonic way.
I think, it is good for you to proceed.

FlorentinBulotAQ replied:
Ok, I will do it

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+
+    def set_to_wakeup(self):
+        self.set_to_sleep(False)
+        return
```
I suggest to use `return True` or to delete the `return`

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+
+        self.serial.flush()
+        time.sleep(2)  # number not specified in datasheet but sensor does not receive command for 2s.
+        return
```
I suggest to use `return True` or to delete the `return`

FlorentinBulotAQ replied:
I would suggest to add some logging, for example something like:

```
if to_sleep:
    self.logger.info("Sensor set to sleep")
else:
    self.logger.info("Sensor set to wakeup - 30 seconds required to obtain stable data")
```

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -18,6 +18,10 @@
+PMS_PASSIVE_MODE = 0
+PMS_ACTIVE_MODE = 1
+
+
```
I would suggest to add the commands as global variables for consistency

```PMS_CMD_CHANGEMODE_PASSIVE = b'\x42\x4d\xe1\x00\x00\x01\x70'
PMS_CMD_CHANGEMODE_ACTIVE = b'\x42\x4d\xe1\x00\x01\x01\x71'
PMS_CMD_READ_IN_PASSIVE = b'\x42\x4d\xe2\x00\x00\x01\x71'
PMS_CMD_TO_SLEEP = b'\x42\x4d\xe4\x00\x00\x01\x73'
PMS_CMD_TO_WAKEUP = b'\x42\x4d\xe4\x00\x01\x01\x74'
```

serunis replied:
ok

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+        self.serial.flush()
+        ret = self.read(False)
+        time.sleep(0.5)  # wait sensor busy finished
+        return ret
```
I suggest to add some logging, for example:

```
if mode == PMS_PASSIVE_MODE:
    self.logger.info("Sensor set in passive mode")
else:
   self.logger.info("Sensor set in active mode")
```
```

serunis replied:
ok

FlorentinBulotAQ commented on plantower/plantower.py:
```diff
@@ -134,3 +138,88 @@ def read(self, perform_flush=True):
+        return
+
+    def read_in_passive(self, perform_flush=True):
+        if perform_flush:
```
Could you add a short docstring at the beginning of this method?

serunis replied:
ok

FlorentinBulotAQ reviewed (commented):
Dear @serunis,

I have tested your code and it works fine on a PMS5003. It was easy to read and to understand and all the methods you have implemented work.
I have just suggested a few minor correction for consistency with the rest of the code. But these are just suggestions so let me know what you think of them, I am happy to discuss it.


The only surprising behaviour is when the sensor is set to sleep, you can still read it using read_in_passive(). I think that is an issue with the sensor and not with the code, please see below:

```
In [15]: pms.set_to_sleep()

In [16]: pms.read()
---------------------------------------------------------------------------
PlantowerException                        Traceback (most recent call last)
<ipython-input-16-fc3fdc851aca> in <module>()
----> 1 pms.read()

/home/pi/plantower/plantower/plantower.py in read(self, perform_flush)
    138                     return PlantowerReading(recv) # convert to reading object
    139             #If the character isn't what we are expecting loop until timeout
--> 140         raise PlantowerException("No message recieved")
    141
    142     def mode_change(self, mode=PMS_PASSIVE_MODE):

PlantowerException: No message recieved
```

```
In [18]: print(pms.read_in_passive())
2020-01-21 11:59:25,5,5,6,6,6,6,948,248,31,2,0,0,

In [19]: print(pms.read_in_passive())
2020-01-21 11:59:29,5,5,6,6,6,6,948,248,31,2,0,0,

In [20]: print(pms.read_in_passive())
2020-01-21 11:59:32,5,5,6,6,6,6,948,248,31,2,0,0,

In [21]: print(pms.read_in_passive())
2020-01-21 11:59:36,5,5,6,6,6,6,948,248,31,2,0,0,

In [22]: print(pms.read_in_passive())
2020-01-21 11:59:39,5,5,6,6,6,6,948,248,31,2,0,0,

```

The buffer does not seem to be cleared and the sensor keeps returning the same data. Something to keep in mind when using the sensor.

Again thanks a lot for your pull request, all these changes are very to the point and very usefull to improve this library.

With best wishes,
Florentin.

serunis committed:
add comment, code style fix

serunis commented:
Hello @FlorentinBulotUoS,
I updated code by your comment.
The usage of self.serial.flush() is intentional. I belive, it makes cpu to wait until tx finished.
and I think it is good for you to make test codes. I do not have a good idea of building test codes.

serunis commented:
>The only surprising behaviour is when the sensor is set to sleep, you can still read it using read_in_passive().

Oh.. I did not know that, funny.

When I measure the power consumption of the sensor in sleep, it was way high for the controller in sleep.
I believe, the sleep mode of the sensor only meant the fan stops.

I originally made my own code for arduino, and I have made a portable outdoor wireless dust sensor solution. however, power comsumption in sleep mode was too higher than I expected, and I change my mind to make it indoor solution with Raspberry pi. 

Thanks to your library, I was happy that I do not have to parse the data structure in python again.
So I commit my impovement to your library.
Since an indoor solution still needs the sensor fan to stop due to its noise, a sleep mode is needed.

Actually this is my first github pull request, quite interesting!

Thank you for review.

FlorentinBulotAQ reviewed (approved):
Thanks a lot for making all these changes!

And I agree with you serial.flush() makes more sense in that case than serial.reset_input_buffer()

FlorentinBulotAQ commented:
Hi @serunis 

I would like to credit you in the README.md file, so far I have added your name as serunis. Let me know if you would like to use a different name.

With best wishes,
Florentin.

serunis commented:
serunis is good for me
Thank you
